rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -0,0 +1,305 @@
+package coop.rchain.rspace
+
+import cats.implicits._
+import com.google.common.collect.Multiset
+import com.typesafe.scalalogging.Logger
+import coop.rchain.catscontrib._
+import coop.rchain.rspace.history.Branch
+import coop.rchain.rspace.internal._
+import coop.rchain.rspace.trace.{Produce, _}
+import coop.rchain.shared.SyncVarOps._
+import scodec.Codec
+
+import scala.Function.const
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Seq
+import scala.concurrent.SyncVar
+import scala.util.Random
+
+class ReplayRSpace[C, P, A, K](val store: IStore[C, P, A, K], val branch: Branch)(
+    implicit
+    serializeC: Serialize[C],
+    serializeP: Serialize[P],
+    serializeA: Serialize[A],
+    serializeK: Serialize[K]
+) extends ISpace[C, P, A, K] {
+
+  private val logger: Logger = Logger[this.type]
+
+  private val replayData: SyncVar[ReplayData] = {
+    val sv = new SyncVar[ReplayData]()
+    sv.put(ReplayData.empty)
+    sv
+  }
+
+  def consume(channels: Seq[C], patterns: Seq[P], continuation: K, persist: Boolean)(
+      implicit m: Match[P, A]): Option[(K, Seq[A])] =
+    store.eventsCounter.registerConsume {
+      if (channels.length =!= patterns.length) {
+        val msg = "channels.length must equal patterns.length"
+        logger.error(msg)
+        throw new IllegalArgumentException(msg)
+      }
+      store.withTxn(store.createTxnWrite()) { txn =>
+        def runMatcher(maybeComm: Option[COMM]): Option[Seq[DataCandidate[C, A]]] = {
+          val channelToIndexedData = channels.map { (c: C) =>
+            c -> {
+              maybeComm match {
+                case Some(comm) =>
+                  store.getData(txn, Seq(c)).zipWithIndex.filter {
+                    case (Datum(_, _, source), _) => comm.produces.contains(source)
+                  }
+                case None =>
+                  Random.shuffle(store.getData(txn, Seq(c)).zipWithIndex)
```

Comment:
You technically don't need to shuffle in this case right? It doesn't hurt to do so but.

Reply:
I guess I thought it was better to mimic the "normal" behavior of `RSpace` in this case.

Reply:
Up to you - it seemed like you wanted to squeeze out every bit in terms of performance so. Maybe just add a comment saying it technically isn't necessary if you decide to keep it?